### ISSUE TITLE
TD012 & TD015 - TargetEndpoint HTTPTargetConnection should have exactly one SSLInfo & Missing MaxFailures element within LoadBalancer

### ISSUE THREAD

**Issue by PratMoha** [apigee/apigeelint#526]
Hi Team,

I'm facing 2 warnings (screenshot below) from Apigee linter which seems to be false positive.

![Image](https://github.com/user-attachments/assets/d836d636-6301-42a0-93fd-19394351f70b)

We usually define the SSLInfo in the edge.json file in the target server section, hence the same is not declared again in the load balancer segment. Is it logically correct for the linter to expect the value for SSLInfo in the target connection?

On the 2nd warning, it seems linter is highlighting the "Missing MaxFailures element within LoadBalancer", but based on the anti-pattern scenarios best practices(point number 3 https://cloud.google.com/apigee/docs/api-platform/antipatterns/load-balancing-maxfailures) 

```
If there's some constraint such that only one TargetServer and if the HealthMonitor is not used, then don't specify MaxFailures in the LoadBalancer configuration.

The default value of MaxFailures is 0. This means that Apigee always tries to connect to the target for each request and never removes the target server from the rotation.
```
Could we add a use-case that if there is only one server in the load balancer then max-failure warning should be supressed?

Regards,
Pratyush

**Comment by ssvaidyanathan** [comment]
@PratMoha - can you share the TargetEndpoint config used in your proxy thats being linted here?

**Comment by PratMoha** [comment]
@ssvaidyanathan ,

Thank you for your swift response. Below is one of the example on how I'm defining the HTTPTargetConnection here server mocktarget-apigee is defined in edge.json inside target server array as defined below.

```
<HTTPTargetConnection>
        <Properties>
            <Property name="success.codes">1XX,2XX,3XX</Property>
        </Properties>
        <LoadBalancer>
            <Server name="mocktarget-apigee"/>
        </LoadBalancer>
    </HTTPTargetConnection>
```

```
"targetServers": [
        {
          "name": "mocktarget-apigee",
          "host": "mocktarget.apigee.net",
          "isEnabled": true,
          "port": 443,
          "sSLInfo": {
            "enabled": "true",
            "enforce": "true"
          }
        }
      ]
```

**Comment by kacperillo** [comment]
Hello, I just have exactly same issue:

1. This is my TargetEndpoint configuration:
`<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<TargetEndpoint name="default">
  <HTTPTargetConnection>
    <LoadBalancer>
      <Server name="PETSTORE" />
    </LoadBalancer>
    <Path>{propertyset.paths.petstore}</Path>
  </HTTPTargetConnection>
</TargetEndpoint>`

I get the linting warning: "**TargetEndpoint HTTPTargetConnection should have exactly one SSLInfo**".
So I added SSLInfo to the configuration:
 `<TargetEndpoint name="default">
    <HTTPTargetConnection>
      **<SSLInfo>
        <Enabled>true</Enabled>
    </SSLInfo>**
    <LoadBalancer>
      <Server name="PETSTORE" />
    </LoadBalancer>
    <Path>{propertyset.paths.petstore}</Path>
  </HTTPTargetConnection>
</TargetEndpoint>`

And then I get another warning: "**When using a LoadBalancer, the SSLInfo should be configured in the TargetServer, not under HTTPTargetConnection**", which is in contradiction to the first one. So in each situation there is a warning. 


2. Regarding the "Missing MaxFailures element within LoadBalancer" warning I confirm - I have only one server specified in <LoadBalancer> and the apigeelint requires <MaxFailures> to be included, however the Google documentation defines it as antipatern: "_Having single TargetServer in a LoadBalancer configuration of the TargetEndpoint or Service Callout policy with MaxFailures set to a non-zero value is not recommended as it can have adverse implications._"

**Comment by DinoChiesa** [comment]
> And then I get another warning: "When using a LoadBalancer, the SSLInfo should be configured in the TargetServer, not under HTTPTargetConnection", which is in contradiction to the first one. So in each situation there is a warning.

ok that's for sure wrong.  I'll have a look.

>  "Missing MaxFailures element within LoadBalancer" warning I confirm - I have only one server specified in LoadBalancer and the apigeelint requires MaxFailures to be included, however the Google documentation defines it as antipatern: "Having single TargetServer in a LoadBalancer configuration of the TargetEndpoint or Service Callout policy with MaxFailures set to a non-zero value is not recommended as it can have adverse implications."

Yes, I'm inclined to keep this warning, as it is an anti-pattern.

WDYT?

**Comment by PratMoha** [comment]
Hi @DinoChiesa,

For TD012, this also gets flagged in Advanced API Security to add a SSLInfo section to the target server, even when the same is defined in the target connection.

For TD015, as per the best practice for anti-pattern mention in the docs https://cloud.google.com/apigee/docs/api-platform/antipatterns/load-balancing-maxfailures

```
If there's some constraint such that only one TargetServer and if the HealthMonitor is not used, then don't specify MaxFailures in the LoadBalancer configuration.

The default value of MaxFailures is 0. This means that Apigee always tries to connect to the target for each request and never removes the target server from the rotation.

```

I have seen a common practice of defining a single server being used by many of my clients as they have a load balancer behind Apigee.

Kindly give your suggestions.
